src/template/diagnostic.rs

```diff
@@ -0,0 +1,424 @@
+//! The code for error/diagnostics handling is heavily inspiered by
```

Comment:
According to the MIT license, we need to put the license text itself here:

> The above copyright notice and this permission notice
> **shall be included** in all copies or substantial portions
> of the Software.

Reply:
Well maybe it's enough if we do it in the copyright notice and reference that from here...

Reply:
I have added a general copyright notice in `src/template/mod.rs`.
I have also added a specific copyright notice above each code item which is inspired by the Rust project with a permalink to the relevant location in rust's source code.

The notices in the source code are implemented as a simple comment (`//`) so that the won't appear in the generated documentation for the crate to avoid cluttering it.
The general notice is a crate documentation (`//!`) and thus will appear in the generated crate documentation.

What do you think about that solution @michidk.

Reply:
Great! Ship it